be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/ParamsWithMapHandler.scala

```diff
@@ -0,0 +1,140 @@
+        Await.ready(ask, duration).value match {
+          case Some(Success(DbActor.DbActorReadChannelDataAck(channelData))) =>
+            val setOfIds = channelData.messages.toSet
+            localHeartBeat += (channel -> setOfIds)
+          case Some(Failure(ex: DbActorNAckException)) => Left(PipelineError(ex.code, s"couldn't readChannelData for local heartbeat", jsonRpcMessage.getId))
+          case reply                                   => Left(PipelineError(ErrorCodes.SERVER_ERROR.id, s"heartbeatHandler failed : unexpected DbActor reply '$reply'", jsonRpcMessage.getId))
+        }
+      })
+
+      /** finally, we only keep from the received heartbeat the message ids that are not contained in the locally extracted heartbeat. * */
+      val missingIds: mutable.HashMap[Channel, Set[Hash]] = mutable.HashMap()
+      receivedHeartBeat.keys.foreach(channel => {
+        if (localHeartBeat.contains(channel)) {
+          missingIds += (channel -> receivedHeartBeat(channel).diff(localHeartBeat(channel)))
+        }
+      })
+      Right(JsonRpcRequest(RpcValidator.JSON_RPC_VERSION, MethodType.GET_MESSAGES_BY_ID, GetMessagesById(missingIds.toMap), None))
+
+    case Right(jsonRpcMessage: JsonRpcResponse) =>
+      Left(PipelineError(ErrorCodes.SERVER_ERROR.id, "HeartbeatHandler received a 'JsonRpcResponse'", jsonRpcMessage.id))
+    case graphMessage @ _ => graphMessage
+  }).filter(!isEmptyAnswer(_))
+
+  private def getMessagesByIdHandler(dbActorRef: AskableActorRef): Flow[GraphMessage, GraphMessage, NotUsed] = Flow[GraphMessage].map({
+    case Right(jsonRpcMessage: JsonRpcRequest) =>
+      val receivedRequest: Map[Channel, Set[Hash]] = jsonRpcMessage.getParams.asInstanceOf[GetMessagesById].channelsToMessageIds
+      val response: mutable.HashMap[Channel, Set[Message]] = mutable.HashMap()
+      receivedRequest.keys.foreach(channel => {
+        val ask = dbActorRef ? DbActor.Catchup(channel)
+        Await.ready(ask, duration).value match {
+          case Some(Success(DbActor.DbActorCatchupAck(messages))) =>
+            val missingMessages = messages.filter(message => receivedRequest(channel).contains(message.message_id)).toSet
+            response += (channel -> missingMessages)
+          case Some(Failure(ex: DbActorNAckException)) => Left(PipelineError(ex.code, s"getMessagesByIdHandler failed : ${ex.message}", jsonRpcMessage.getId))
+          case reply                                   => Left(PipelineError(ErrorCodes.SERVER_ERROR.id, s"AnswerGenerator failed : unexpected DbActor reply '$reply'", jsonRpcMessage.getId))
+        }
+      })
+      Right(JsonRpcResponse(RpcValidator.JSON_RPC_VERSION, new ResultObject(HashMap.from(response)), None))
+
+    case Right(jsonRpcMessage: JsonRpcResponse) =>
+      Left(PipelineError(ErrorCodes.SERVER_ERROR.id, "getMessagesByIdHandler received a 'JsonRpcResponse'", jsonRpcMessage.id))
+    case graphMessage @ _ => graphMessage
+  })
```

Comment:
no need for the closing parenthesis
```suggestion
  }
```